File: day11/part1.rb
require "set"

Position = Struct.new(:i, :j) do
  def neighbors(existing_positions)
    [
      Position.new(i, j + 1),
      Position.new(i, j - 1),
      Position.new(i + 1, j),
      Position.new(i - 1, j),
      Position.new(i + 1, j + 1),
      Position.new(i + 1, j - 1),
      Position.new(i - 1, j + 1),
      Position.new(i - 1, j - 1),
    ].select { |c| existing_positions.member?(c) }
  end
end

class Octopus
  attr_accessor :position
  attr_accessor :last_flash_step

  def initialize(energy_level)
    @energy_level = energy_level
    @last_fash_step = -1
  end

  def increase_energy_level!(for_step_number)
    if @last_fash_step < for_step_number
      @energy_level += 1
    end

    if @energy_level > 9
      @last_fash_step = for_step_number
      @energy_level = 0
      return true
    end

    false
  end
end

class Simulation
  attr_reader :flash_count

  def initialize
    @flash_count = 0
    @step_number = 0
    @octopuses = {}
    @positions = Set.new
    ARGF.each_line.each.with_index do |line, i|
      line.chomp.each_char.with_index do |energy_level, j|
        position = Position.new(i, j)
        @octopuses[position] = Octopus.new(energy_level.to_i)
        @positions << position
      end
    end
  end

  def tick
    @step_number += 1
    @positions.each do |position|
      increase_and_propagate_flashes!(position)
    end
  end

  private

  def increase_and_propagate_flashes!(position)
    positions_to_increase = [position]
    while positions_to_increase.any?
      position_to_increase = positions_to_increase.shift
      do_flash = @octopuses
        .fetch(position_to_increase)
        .increase_energy_level!(@step_number)
      if do_flash
        @flash_count += 1
        positions_to_increase.concat(
          position_to_increase.neighbors(@positions)
        )
      end
    end
  end
end

if __FILE__ == $0
  simulation = Simulation.new
  100.times { simulation.tick }
  puts(simulation.flash_count)
end

File: day11/part1.py
from dataclasses import dataclass
from typing import Callable, Container, Dict, Iterable

@dataclass(frozen=True)
class Position:
    i: int
    j: int

@dataclass(frozen=True)
class Octopus:
    energy_level: int

OctopusesMap = Dict[Position, Octopus]

def neighbors(positions: Container[Position], position: Position) -> Iterable[Position]:
    return filter(
        lambda p: p in positions,
        [
            Position(position.i, position.j + 1),
            Position(position.i, position.j - 1),
            Position(position.i + 1, position.j),
            Position(position.i - 1, position.j),
            Position(position.i + 1, position.j + 1),
            Position(position.i + 1, position.j - 1),
            Position(position.i - 1, position.j + 1),
            Position(position.i - 1, position.j - 1),
        ]
    )

def step(octopuses: OctopusesMap, on_flash: Callable) -> OctopusesMap:
    octopuses = octopuses.copy()
    positions_to_reset = set()

    # Increase and propagate flashes for each positions
    for position in octopuses.keys():
        positions_to_increase = [position]
        while positions_to_increase:
            position_to_increase, *positions_to_increase = positions_to_increase
            did_flash = _increase_energy_level(octopuses, position_to_increase)
            if did_flash:
                positions_to_reset.add(position_to_increase)
                neighbors_to_increase = neighbors(positions=octopuses, position=position_to_increase)
                positions_to_increase.extend(neighbors_to_increase)

                if on_flash:
                    on_flash()

    # Reset the octopus that flashed after a step
    for position_to_reset in positions_to_reset:
        octopuses[position_to_reset] = Octopus(energy_level=0)

    return octopuses

def _increase_energy_level(
    octopuses: OctopusesMap,
    position: Position,
    max_energy_level: int = 10,
) -> bool:
    octopus = octopuses[position]
    if octopus.energy_level < max_energy_level:
        octopuses[position] = Octopus(octopus.energy_level + 1)
        return octopus.energy_level + 1 == max_energy_level
    return False

if __name__ == "__main__":
    import fileinput

    octopuses = {
        Position(i, j): Octopus(energy_level=int(cell))
        for i, line in enumerate(fileinput.input())
        for j, cell in enumerate(line.strip())
    }

    flash_count = 0
    def on_flash():
        nonlocal flash_count
        flash_count += 1

    for _ in range(100):
        octopuses = step(octopuses, on_flash)

    print(flash_count)
